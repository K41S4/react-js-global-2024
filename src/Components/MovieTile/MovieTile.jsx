import { Link, useSearchParams } from 'react-router-dom';
import { GENRE_CONNECTOR } from '../../constants';
import styles from './MovieTile.module.css';

export const MovieTile = ({ id, imageUrl, movieName, releaseYear, relevantGenres }) => {
  const [searchParams] = useSearchParams();
  return (
    <Link className={styles.link} to={`/${id}?${searchParams.toString()}`} data-cy="movie-tile">
      <div className={styles.container}>
        <img className={styles.image} src={imageUrl} alt="" />
        <div className={styles.detailsContainer}>
          <div className={styles.movieName}>{movieName}</div>
          <div className={styles.releaseYear}>{releaseYear}</div>
        </div>
        {!!relevantGenres?.length && (
          <div className={styles.relevantGenres}>{relevantGenres.join(GENRE_CONNECTOR)}</div>
        )}
      </div>
    </Link>
  );
};
