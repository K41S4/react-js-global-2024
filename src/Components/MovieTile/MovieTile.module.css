.container {
  width: 250px;
  min-width: 250px;
  color: white;
  background-color: transparent;
  overflow: hidden;
  font-weight: 100;
  font-family: 'Segoe UI', sans-serif;
}

.image {
  width: 100%;
  display: block;
}

.movieName {
  font-size: 18px;
  white-space: normal;
  overflow: visible;
  display: inline-block;
}

.detailsContainer {
  display: flex;
  justify-content: space-between;
  padding: 8px;
}

.releaseYear {
  display: inline-block;
  min-width: 80px;
  height: 26px;
  line-height: 26px;
  border-radius: 10%;
  border: 1px solid #979797;
  text-align: center;
  font-size: 14px;
  background-color: transparent;
}

.relevantGenres {
  font-size: 14px;
  color: #979797;
  padding: 8px;
}
