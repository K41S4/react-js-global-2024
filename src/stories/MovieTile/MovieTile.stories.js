import { fn } from '@storybook/test';
import { MovieTile } from '../../Components/MovieTile/MovieTile';

export default {
  title: 'Example/MovieTile',
  component: MovieTile,
  parameters: {
    layout: 'centered',
    backgrounds: {
      default: 'dark',
    },
  },
  args: {
    onClick: fn(),
  },
  tags: ['autodocs'],
};

export const MovieTileDefault = {
  args: {
    imageUrl:
      'https://m.media-amazon.com/images/M/MV5BYTViYTE3ZGQtNDBlMC00ZTAyLTkyODMtZGRiZDg0MjA2YThkXkEyXkFqcGc@._V1_.jpg',
    movieName: 'Pulp Fiction',
    releaseYear: '1999',
    relevantGenres: ['Crime', 'Comedy'],
  },
};
