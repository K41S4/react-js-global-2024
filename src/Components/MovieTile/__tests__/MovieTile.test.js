import { screen, waitFor } from '@testing-library/react';
import { MovieTile } from '../MovieTile';
import userEvent from '@testing-library/user-event';
import { Route, Routes } from 'react-router-dom';
import { renderWithRouter } from '../../../testUtils';

describe('MovieTile', () => {
  test('renders all details', () => {
    const testData = {
      id: '123',
      imageUrl: 'imageUrl',
      title: 'testName',
      releaseYear: 'testYear',
      relevantGenres: ['genre1', 'genre2'],
    };

    const { asFragment } = renderWithRouter(<MovieTile {...testData} />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('renders with required props', () => {
    const testData = {
      id: '123',
      imageUrl: 'imageUrl',
      title: 'testName',
      releaseYear: 'testYear',
    };

    const { asFragment } = renderWithRouter(<MovieTile {...testData} />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('navigates to movie id url on container click', async () => {
    const defaultProps = {
      id: '123',
      imageUrl: 'imageUrl',
      title: 'testName',
      releaseYear: 'testYear',
      relevantGenres: ['genre1', 'genre2'],
    };

    const user = userEvent.setup();

    renderWithRouter(
      <Routes>
        <Route path="/" element={<MovieTile {...defaultProps} />} />
        <Route path="/123" element={<h1>Movie Page</h1>} />
      </Routes>
    );

    await user.click(screen.getByRole('link'));

    expect(screen.getByText('Movie Page')).toBeInTheDocument();
  });

  test('opens context menu on container right click', async () => {
    const defaultProps = {
      id: '123',
      imageUrl: 'imageUrl',
      title: 'testName',
      releaseYear: 'testYear',
      relevantGenres: ['genre1', 'genre2'],
    };

    const user = userEvent.setup();

    const { container } = renderWithRouter(<MovieTile {...defaultProps} />);

    await user.pointer(
      { keys: '[MouseRight]', target: container.firstElementChild },
      { keys: '[/MouseRight]', target: container.firstElementChild }
    );

    waitFor(() => expect(screen.getByText('Edit')).toBeInTheDocument());
  });
});
