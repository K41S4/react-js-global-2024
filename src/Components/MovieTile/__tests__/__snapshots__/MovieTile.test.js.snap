// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieTile renders all details 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <a
      class="link"
      data-cy="movie-tile"
      href="/123"
    >
      <img
        alt=""
        class="image"
        src="imageUrl"
      />
      <div
        class="detailsContainer"
      >
        <div
          class="title"
        >
          testName
        </div>
        <div
          class="releaseYear"
        >
          testYear
        </div>
      </div>
      <div
        class="relevantGenres"
      >
        genre1 & genre2
      </div>
    </a>
  </div>
</DocumentFragment>
`;

exports[`MovieTile renders with required props 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <a
      class="link"
      data-cy="movie-tile"
      href="/123"
    >
      <img
        alt=""
        class="image"
        src="imageUrl"
      />
      <div
        class="detailsContainer"
      >
        <div
          class="title"
        >
          testName
        </div>
        <div
          class="releaseYear"
        >
          testYear
        </div>
      </div>
    </a>
  </div>
</DocumentFragment>
`;
